indexer/matcher.hpp

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include "indexer/feature_data.hpp"
+#include "indexer/ftypes_mapping.hpp"
+
+#include "coding/text_file_reader.hpp"
+
+#include "base/assert.hpp"
+#include "base/logging.hpp"
+
+#include <cstdint>
+#include <initializer_list>
+#include <string>
+
+namespace matcher
+{
+template<typename Base, typename Type, Type notFound>
+class Matcher
+{
+public:
+  static Type GetType(feature::TypesHolder const & types)
+  {
+    static Base instance;
+    auto const it = instance.m_matcher.Find(types);
+    if (!instance.m_matcher.IsValid(it))
+      return notFound;
+
+    return it->second;
+  }
+protected:
+  ftypes::HashMapMatcher<uint32_t, Type> m_matcher;
+};
+
+enum UGCType
+{
+  None    = 0,
+  Rating  = 1 << 0, // 1
+  Reviews = 1 << 1, // 2
+  Details = 1 << 2  // 4
+};
+
+inline UGCType operator|(UGCType lhs, UGCType rhs)
+{
+  return static_cast<UGCType>(static_cast<int>(lhs) | static_cast<int>(rhs));
+}
+
+class UGCMatcher: public Matcher<UGCMatcher, UGCType, None>
+{
+  friend class Matcher;
+
+  UGCMatcher()
+  {
+    text_file_reader::ReadLineByLine("ugc_types.csv", [this](std::vector<std::string> const & types,
+                                                        std::vector<std::string> const & mask)
+                                    {
```

Comment:
This lambda looks ugly - lots of useless indents. Is it possible to reformat it or move to an individual function?